File: rtls_flutter/ios/Classes/RtlsFlutterPlugin.swift
import Flutter
import Foundation
import RTLSyncKit

public class RtlsFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private var client: LocationSyncClient?
    private var eventSink: FlutterEventSink?
    private var eventTask: Task<Void, Never>?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.rtls.flutter/rtls", binaryMessenger: registrar.messenger())
        let instance = RtlsFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        let eventChannel = FlutterEventChannel(name: "com.rtls.flutter/rtls_events", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":
            guard let args = call.arguments as? [String: Any],
                  let baseURLStr = args["baseUrl"] as? String,
                  let baseURL = URL(string: baseURLStr),
                  let userId = args["userId"] as? String,
                  let deviceId = args["deviceId"] as? String,
                  let accessToken = args["accessToken"] as? String else {
                result(FlutterError(code: "INVALID", message: "baseUrl, userId, deviceId, accessToken required", details: nil))
                return
            }
            let token = accessToken
            let provider = AuthTokenProvider { token }
            let databaseURL: URL
            do {
                databaseURL = try LocationSyncClientConfiguration.defaultDatabaseURL(directoryName: "RTLSyncKitFlutter")
            } catch {
                result(FlutterError(code: "CONFIG", message: error.localizedDescription, details: nil))
                return
            }
            let config = LocationSyncClientConfiguration(
                baseURL: baseURL,
                authTokenProvider: provider,
                userId: userId,
                deviceId: deviceId,
                databaseURL: databaseURL
            )
            Task {
                do {
                    let c = try await LocationSyncClient(configuration: config)
                    await MainActor.run {
                        self.client = c
                        self.startEventLoop(client: c)
                        result(nil)
                    }
                } catch {
                    await MainActor.run {
                        result(FlutterError(code: "CONFIG", message: error.localizedDescription, details: nil))
                    }
                }
            }
        case "startTracking":
            guard let client = client else {
                result(FlutterError(code: "NOT_CONFIGURED", message: "Call configure first", details: nil))
                return
            }
            Task {
                await client.startTracking()
                await MainActor.run { result(nil) }
            }
        case "stopTracking":
            Task {
                await client?.stopTracking()
                await MainActor.run { result(nil) }
            }
        case "requestAlwaysAuthorization":
            Task {
                await client?.requestAlwaysAuthorization()
                await MainActor.run { result(nil) }
            }
        case "getStats":
            guard let client = client else {
                result(["pendingCount": -1, "oldestPendingRecordedAtMs": NSNull()])
                return
            }
            Task {
                let stats = await client.stats()
                let oldest: Any = stats.oldestPendingRecordedAt.map { Int($0.timeIntervalSince1970 * 1000) } as Any? ?? NSNull()
                await MainActor.run {
                    result(["pendingCount": stats.pendingCount, "oldestPendingRecordedAtMs": oldest])
                }
            }
        case "flushNow":
            Task {
                await client?.flushNow()
                await MainActor.run { result(nil) }
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func startEventLoop(client: LocationSyncClient) {
        eventTask?.cancel()
        eventTask = Task { [weak self] in
            guard let self = self else { return }
            for await event in client.events {
                if Task.isCancelled { break }
                await MainActor.run {
                    self.eventSink?(self.eventToMap(event))
                }
            }
        }
    }

    private func eventToMap(_ event: LocationSyncClientEvent) -> [String: Any] {
        switch event {
        case .recorded(let point):
            return [
                "type": "recorded",
                "point": [
                    "id": point.id.uuidString,
                    "userId": point.userId,
                    "deviceId": point.deviceId,
                    "recordedAtMs": Int(point.recordedAt.timeIntervalSince1970 * 1000),
                    "lat": point.coordinate.latitude,
                    "lng": point.coordinate.longitude
                ]
            ]
        case .syncEvent(let e):
            switch e {
            case .didUpload(let accepted, let rejected):
                return ["type": "syncEvent", "event": "uploadSucceeded", "accepted": accepted, "rejected": rejected]
            case .uploadFailed(let msg):
                return ["type": "syncEvent", "event": "uploadFailed", "message": msg]
            }
        case .error(let msg):
            return ["type": "error", "message": msg]
        case .trackingStarted:
            return ["type": "trackingStarted"]
        case .trackingStopped:
            return ["type": "trackingStopped"]
        case .authorizationChanged:
            return ["type": "authorizationChanged"]
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: rtls_flutter/android/src/main/kotlin/com/rtls/rtls_flutter/RtlsFlutterPlugin.kt
package com.rtls.rtls_flutter

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import com.rtls.kmp.LocationSyncClientEvent
import com.rtls.kmp.RTLSKmp

class RtlsFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private var channel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var client: com.rtls.kmp.LocationSyncClient? = null
    private var context: Context? = null
    private var lastUserId: String? = null
    private var lastDeviceId: String? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "com.rtls.flutter/rtls")
        channel!!.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "com.rtls.flutter/rtls_events")
        eventChannel!!.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        client?.stopTracking()
        client = null
        context = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "configure" -> {
                val baseUrl = call.argument<String>("baseUrl") ?: ""
                val userId = call.argument<String>("userId") ?: ""
                val deviceId = call.argument<String>("deviceId") ?: ""
                val accessToken = call.argument<String>("accessToken") ?: ""
                val ctx = context ?: run {
                    result.error("NO_CONTEXT", "Context not available", null)
                    return
                }
                client?.stopTracking()
                lastUserId = userId
                lastDeviceId = deviceId
                client = RTLSKmp.createLocationSyncClient(ctx, baseUrl, userId, deviceId, accessToken, scope)
                result.success(null)
            }
            "startTracking" -> {
                val ctx = context ?: run {
                    result.error("NO_CONTEXT", "Context not available", null)
                    return
                }
                val c = client
                if (c == null) {
                    result.error("NOT_CONFIGURED", "Call configure first", null)
                    return
                }
                val userId = lastUserId ?: return result.error("INVALID", "Configure first", null)
                val deviceId = lastDeviceId ?: return result.error("INVALID", "Configure first", null)
                val flow = RTLSKmp.createLocationFlow(ctx, userId, deviceId)
                c.startCollectingLocation(flow)
                result.success(null)
            }
            "stopTracking" -> {
                client?.stopTracking()
                result.success(null)
            }
            "requestAlwaysAuthorization" -> result.success(null)
            "getStats" -> {
                scope.launch {
                    val stats = try {
                        client?.stats()
                    } catch (e: Exception) {
                        null
                    }
                    withContext(Dispatchers.Main) {
                        if (stats != null) {
                            result.success(mapOf(
                                "pendingCount" to stats.pendingCount,
                                "oldestPendingRecordedAtMs" to (stats.oldestPendingRecordedAtMs ?: 0)
                            ))
                        } else {
                            result.success(mapOf("pendingCount" to -1, "oldestPendingRecordedAtMs" to null))
                        }
                    }
                }
            }
            "flushNow" -> {
                scope.launch {
                    try {
                        client?.flushNow()
                        withContext(Dispatchers.Main) { result.success(null) }
                    } catch (e: Exception) {
                        withContext(Dispatchers.Main) { result.error("FLUSH_FAILED", e.message, null) }
                    }
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        scope.launch {
            client?.events?.collectLatest { event ->
                val map = when (event) {
                    is LocationSyncClientEvent.Recorded -> mapOf(
                        "type" to "recorded",
                        "point" to mapOf(
                            "id" to event.point.id,
                            "userId" to event.point.userId,
                            "deviceId" to event.point.deviceId,
                            "recordedAtMs" to event.point.recordedAtMs,
                            "lat" to event.point.lat,
                            "lng" to event.point.lng
                        )
                    )
                    is LocationSyncClientEvent.SyncEvent -> mapOf(
                        "type" to "syncEvent",
                        "event" to when (val e = event.event) {
                            is com.rtls.kmp.SyncEngineEvent.UploadSuccess -> "uploadSucceeded"
                            is com.rtls.kmp.SyncEngineEvent.UploadFailed -> "uploadFailed"
                            else -> "unknown"
                        }
                    )
                    is LocationSyncClientEvent.Error -> mapOf("type" to "error", "message" to event.message)
                    LocationSyncClientEvent.TrackingStarted -> mapOf("type" to "trackingStarted")
                    LocationSyncClientEvent.TrackingStopped -> mapOf("type" to "trackingStopped")
                }
                kotlinx.coroutines.withContext(Dispatchers.Main) { eventSink?.success(map) }
            }
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}
